sdk/python/lib/pulumi/provider/server.py

```diff
@@ -158,6 +171,86 @@ async def _construct_response(self, result: ConstructResult) -> proto.ConstructR
                                        state=state,
                                        stateDependencies=deps)
 
+
+    async def Call(self, request: proto.CallRequest, context):  # pylint: disable=invalid-overridden-method
+        # Calls to `Construct` and `Call` are serialized because they currently modify globals. When we are able to
+        # avoid modifying globals, we can remove the locking.
+        await self.lock.acquire()
+        try:
+            return await self._call(request, context)
+        finally:
+            self.lock.release()
+
+    async def _call(self, request: proto.CallRequest, context):
+        # pylint: disable=unused-argument
+        assert isinstance(request, proto.CallRequest), \
+            f'request is not CallRequest but is {type(request)} instead'
+
+        pulumi.runtime.settings.reset_options(
```

Comment:
As before, we don't reset them back to original values do we, when done? That is nothing like:

```
with temp_runtime_settings(...):
    ....
```

Perhaps this is fine..

Also would we be interested in removing the lock actually?

1. the best would be to get to explicit parameter passing; perhaps a lot of work/ breaking?

2. we could try to use https://www.python.org/dev/peps/pep-0555/#semantics-and-higher-level-api coroutine-scoped context vars;  the change probably would be lexically very small, however I'd want to personally test out how context vars behave to confirm the assumptions before shipping something like that, it sounds a bit risky

Reply:
We'll want to remove the lock eventually.